Fixes #21287 - add resource switcher

amirfefer opened the pull request.

waldenraines commented:
@amirfefer how do you feel about deploying the storybook so I can have a look more easily?

amirfefer commented:
@waldenraines  sure :+1:

timogoebel commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.stories.js:
```diff
@@ -0,0 +1,23 @@
+import { storiesOf } from '@storybook/react';
+import ResourceSwitcher from './';
+
+storiesOf('Resource Switcher', module).add('swithcer', () => (
```
Nit: should be `switcher`

amirfefer commented:
@waldenraines 
Storybook link: http://rawgit.com/amirfefer/foreman/resource-switcher/index.html?selectedKind=Resource%20Switcher&selectedStory=swithcer&full=0&down=1&left=1&panelRight=0&downPanel=storybook%2Factions%2Factions-panel

amirfefer commented:
@sharvit  could you have a look?

sharvit commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.css:
```diff
@@ -0,0 +1,5 @@
+.scrollable-list {
+  max-height: 300px;
```
Would be nice to use variables here, so plugins could override them instead of overriding the class itself.

sharvit commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResurceSwitcher.test.js:
```diff
@@ -0,0 +1,15 @@
+import { resourcesProp } from './ResourceSwitcher.fixtures';
+import ResourceSwitcher from './';
+
+jest.unmock('./');
```
Really need it here? Shouldn't we allow jest to mock all the imports?

sharvit commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js:
```diff
@@ -0,0 +1,37 @@
+    <Popover id="popover-contained" title={__('Resource Switcher')}>
+      <ListGroup className="scrollable-list">
+        {resources.map(item => (
+          <ListGroupItem key={`id-${item.caption}`} href={item.url}>
```
Would it make sense to support both `href` and `onClick`?

sharvit commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js:
```diff
@@ -0,0 +1,37 @@
+  );
+};
+
+export default ResourceSwitcher;
```
Can you add `propTypes` please?

sharvit reviewed (changes requested):
Overall, it feels good.

1. Is it still WIP? if so, what is actually missing here?

2. You can use the `onOutsideClick` to allow auto toggling:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/notifications/index.js#L80

3. The design doesn't look like the patternfly design.

amirfefer commented:
@sharvit 
Thanks for your review!
I've removed `WIP` tag from this PR.
I addressed your inline comments plus,
Instead of using `onOutsideClick` as you suggested, I used `rootClose` prop (already implemented in react-bootstrap)

waldenraines commented on webpack/assets/javascripts/react_app/components/bookmarks/__snapshots__/bookmarks.test.js.snap:
```diff
@@ -1906,4 +1906,4 @@ exports[`bookmarks should show no bookmarks if server did not respond with any 1
   </Connect(BookmarkContainer)>
 </Provider>
-`;
+`;
```
Nitpick, seems accidental.

waldenraines commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js:
```diff
@@ -0,0 +1,57 @@
+
+const ResourceSwitcher = ({ resources }) => {
+  const popover = (
+    <Popover id="popover-contained" title={__('Resource Switcher')}>
```
What if we just omit the title?  If you look at the version in katello it doesn't have a title but instead has a search bar.

<img width="534" alt="screen shot 2018-02-20 at 2 25 44 pm" src="https://user-images.githubusercontent.com/4116405/36444670-f7fcdce4-1649-11e8-8ca9-756475bfa19c.png">

sharvit reviewed (changes requested):
LGTM except for the filename changing.

Do you have to change the filename?

This pr #5212 is changing the filename same as you do:
https://github.com/theforeman/foreman/pull/5212/files#diff-c81edf5233381be438a6cd30882c7c19R1
https://github.com/theforeman/foreman/pull/5212/files#diff-f394b153b2a3a9f4f039a8743b649144R1

I would prefer you will keep it the same filename so it will be easier for me to rebase.

amirfefer committed:
Fixes #21287 - add resource switcher

amirfefer commented:
@waldenraines Thanks, I removed that title.
@sharvit  - I changed the file name back.

sharvit reviewed (approved):
LGTM!
@ohadlevy

amirfefer commented:
[test katello]

amirfefer commented:
@ohadlevy could we merge this?

waldenraines reviewed (commented):
I hate to bring this up so late but what do we think about calling this "Item Switcher" instead of "Resource Switcher"?  I'm not opposed to "Resource Switcher" but "Item Switcher" may be more generic and is shorter.

Otherwise this looks good to me.

dLobatog reviewed (commented):
Looks good to me too, just 2 questions:

- I see the Katello one is not a button but a clickable link, should this be the same?
- What page is this going to be used on?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.stories.js:
```diff
@@ -0,0 +1,24 @@
+storiesOf('Resource Switcher', module).add('resource switcher', () => (
+  <ResourceSwitcher
+    resources={[
+      { caption: 'Host 1', url: '#', onClick: undefined },
```
just wondering, why not use the fixtures? they seem very close?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js:
```diff
@@ -0,0 +1,57 @@
+  const popover = (
+    <Popover id="popover-contained">
+      <ListGroup className="scrollable-list">
+        {resources.map((item) => {
```
what if this is a very long list? should we limit it somehow? did you guys discuss on if we should filter it?

amirfefer replied:
yes we should use an autocomplete search bar, as soon as `autocomplete` component will merge, I'll update this. With this iterative approach I think we can merge components faster

waldenraines replied:
The original mockups for the katello side include search and pagination of the results.

amirfefer replied:
@waldenraines 
I don't like the pagination here, looks odd in a dropdown menu, patternfly design doesn't include it either.

waldenraines replied:
@Rohoover thoughts?

Rohoover replied:
I included pagination because of the large number of rows. Forcing the user to search for something doesn't allow for browsing. I also did not want to re-introduce anything like infinite scroll since previous customer feedback was disastrous.

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js:
```diff
@@ -0,0 +1,57 @@
+import React from 'react';
```
just wondering, as i didn't fully catched up, is this structure in line with @sharvit proposal ?

amirfefer replied:
no... it's not a redux connected component.

amirfefer commented:
@dLobatog Thanks for your comment.
I followed [patternfly design](http://www.patternfly.org/pattern-library/navigation/breadcrumbs/#design), Patternfly uses a button for it, as you can see here:
![item-switcher](https://user-images.githubusercontent.com/11807069/36730624-ba437dea-1bd0-11e8-9e1a-b57dbbdcff1f.png)
and it should combined with the breadcrumbs, in order to switch to a different item - another way to navigate. 

@waldenraines I don't mind to replace the name, `item switcher` sounds even better for me.

waldenraines commented:
> I see the Katello one is not a button but a clickable link, should this be the same?

~~@Rohoover thoughts on this?  I can update the katello side if we decide on a button.~~

ETA: never mind, I didn't realize that this is now in patternfly.  I will update the katello side to use a button.

waldenraines commented:
> @waldenraines I don't mind to replace the name, item switcher sounds even better for me.

Maybe we should call it breadcrumb switcher to match the name of the patternfly design?  I didn't realize that this had been put into patternfly.  In that case I wonder if it makes sense to move this to patternfly-react.

Rohoover reviewed (commented):
@waldenraines 

It was in process in PF when you implemented. Your implementation I think, quickened its contribution :)
 Sorry I should have notified you.

amirfefer commented:
@waldenraines, I renamed it to `BreadcrumbSwitcher`
@ohadlevy - I reuse `fixtures` file in the storybook story file.

amirfefer commented:
[test foreman]

waldenraines commented:
@amirfefer and what are your thoughts on going ahead and moving this into patternfly-react?

amirfefer commented:
@waldenraines I totally agree on that, though auto complete component still WIP, so it would take some time until this get merge there and the time quite pressing. we can do these both - merge this into foreman, and open a PR in pf-react ?

waldenraines commented:
> @waldenraines I totally agree on that, though auto complete component still WIP, so it would take some time until this get merge there and the time quite pressing. we can do these both - merge this into foreman, and open a PR in pf-react ?

Works for me, thanks @amirfefer

amirfefer commented:
@ohadlevy 
could we merge this?

waldenraines commented on webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/BreadcrumbSwitcher.test.js:
```diff
@@ -0,0 +1,13 @@
+import { shallow } from 'enzyme';
+import React from 'react';
+import { ResourcesProps } from './BreadcrumbSwitcher.fixtures';
+import ResourceSwitcher from './';
```
Doesn't this need to be updated to BreadcrumbSwitcher?

ohadlevy commented:
@amirfefer can you rebase please? tests are currently failing.

ohadlevy commented on .storybook/config.js:
```diff
@@ -1,5 +1,6 @@
 import { configure } from '@storybook/react';
 
+window.jest = { fn: () => {} };
```
are you mocking all window objects?? did you mean to do this instead: https://github.com/theforeman/foreman/pull/5019/files#diff-a7e25ac69ff4b688eb41c907225d770bR19

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/index.js:
```diff
@@ -0,0 +1,57 @@
+            <ListGroupItem
+              className="no-border"
+              key={`id-${caption}`}
+              href={url}
```
will this work with turbolinks?

amirfefer replied:
yes, `ListGroupItem` renders `<a>` tag when href exists.

ohadlevy reviewed (commented):
a couple of in line questions, but a more generic question is why isnt this part of pf-react instead of foreman to begin with? thanks!

ohadlevy commented:
@amirfefer also a rebase is needed... please make sure your PR are green....

amirfefer commented:
@ohadlevy, @waldenraines   thanks
fixed and rebased.

waldenraines commented:
> a couple of in line questions, but a more generic question is why isnt this part of pf-react instead of foreman to begin with? thanks!

As @amirfefer detailed above, we need this for 6.4 and there is some concern that it won't get into patternfly-react in a timely manner.  We fully intend to move this to patternfly-react very soon though.

waldenraines commented:
@amirfefer looks like there are additional changes needed here to get this green.

amirfefer commented:
@waldenraines  tests seems unrelated

ohadlevy commented on .storybook/preview-head.html:
```diff
@@ -10,4 +10,8 @@
     return str;
   }
+  
+  window.jest = {
```
1. why do we need storybook code to be aware of jest? vs a mock in a test?
2. this is pure JS (doesnt go via webpack) hence you might shouldnt use ES6.

amirfefer replied:
because i'm using fixtures file in order to be more dry.
https://github.com/theforeman/foreman/pull/5188/files#diff-7f590775f94b18cdd0bc39583a9a7d40R6

ohadlevy commented:
[test foreman]

ohadlevy commented:
[test katello]

ohadlevy commented:
[test upgrade]

ohadlevy reviewed (commented):
I would be happy to see a PR to patternfly soon, I really can't see a reason why not to merge it to PF

waldenraines commented:
> I would be happy to see a PR to patternfly soon, I really can't see a reason why not to merge it to PF

If I recall correctly it's because we are waiting on finalized design from patternfly.  @Rohoover is there a finalized design for the breadcrumb switcher yet?

Rohoover commented:
@waldenraines 
I believe so... it can't be found here:

https://github.com/Rohoover/patternfly-design/blob/master/pattern-library/navigation/breadcrumbs/design/design.md

Is that what you're looking for?

waldenraines commented:
> @waldenraines I totally agree on that, though auto complete component still WIP, so it would take some time until this get merge there and the time quite pressing. we can do these both - merge this into foreman, and open a PR in pf-react ?

@amirfefer the autocomplete component is one that I know was waiting on a finalized design (see https://github.com/patternfly/patternfly-react/pull/190#issuecomment-364109632) and that was the reason for the delay.  I think we should try getting this into patternfly-react and if there are problems doing so we can move this back into foreman temporarily.

waldenraines commented:
> @amirfefer the autocomplete component is one that I know was waiting on a finalized design (see patternfly/patternfly-react#190 (comment)) and that was the reason for the delay. I think we should try getting this into patternfly-react and if there are problems doing so we can move this back into foreman temporarily.

@amirfefer thoughts on this?  Honestly I've seen patternfly-react be much quicker than getting foreman PRs merged 😄

ohadlevy commented:
thats probably my fault, if you can help me identify which PR's are tested
/ ready to be merged I'll be happy to have another look

On Thu, Mar 22, 2018 at 3:34 PM, Walden Raines <notifications@github.com>
wrote:

> @amirfefer <https://github.com/amirfefer> the autocomplete component is
> one that I know was waiting on a finalized design (see
> patternfly/patternfly-react#190
> <https://github.com/patternfly/patternfly-react/pull/190> (comment)) and
> that was the reason for the delay. I think we should try getting this into
> patternfly-react and if there are problems doing so we can move this back
> into foreman temporarily.
>
> @amirfefer <https://github.com/amirfefer> thoughts on this? Honestly I've
> seen patternfly-react be much quicker than getting foreman PRs merged 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/5188#issuecomment-375307448>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx7HRrE46TrjqClIPNKuhXIfEJtL4ks5tg6hqgaJpZM4RjXqL>
> .
>

waldenraines commented:
> thats probably my fault, if you can help me identify which PR's are tested
> / ready to be merged I'll be happy to have another look

Not your fault just an observation.  It's really the small number of people with commit access that is the problem.  We are working on helping to alleviate this on our team.

waldenraines commented:
@amirfefer was this closed because we are moving this to patternfly-react?
